### DISCUSSION TITLE
Are classes without any fields or instance method treated as a static class ?

### DISCUSSION THREAD

**Discussion by Xyncgas** [dotnet/runtime#66256] [Q&A]
say I have a class that has no fields, no instance methods, only static methods
but I don't make the class a static class because it's an abstract class (you can't mark a class as static abstract in c#)
does .NET still instantiate this class or it just trimmed down to function pointers to these static functions?

**Comment by Clockwork-Muse** [comment]
You can't instantiate it because it's `abstract` (regardless of optimizations), although you could instantiate a subclass.
I'm assuming that's not what you're really asking about, though.

Why do you want to know?  What are you attempting/planning to do with that information?

**Comment by huoyaoyuan** [comment]
Note: in IL metadata and C++/CLI, `static class` is actually encoded as `abstract sealed`.

**Comment by Joe4evr** [comment]
Let me start by pointing out the API design used for hardware intrinsics: https://source.dot.net/#System.Private.CoreLib/Avx.cs
All of these classes contain only `static` members, yet they're not marked `static` but rather `abstract` (and constructors hidden) so that the types can still participate in an appropriate hierarchy (if a CPU knows SSE4.2 instructions then by definition it has to know all prior SSE instructions, AVX2 should know AVX, etc.).

Now, as far as the type system is concerned, these are ***not*** `static` types. It still allows the type to be placed in input/output positions (eg. `M(Avx avx)`/`Avx M()`), but neither of these make any sense since external code can't derive a *concrete* type to instantiate. (Potentially, an analyzer could do some work here, for example applying a new attribute `[StaticHolderType]` (🍝) on these classes would run appropriate checks on the class itself, as well as warning consumers when they use the type in ways that only *instances* would work.)
